Spigot-API-Patches/0285-Custom-Advancement-Triggers.patch

```diff
@@ -0,0 +1,129 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Jake Potrebic <jake.m.potrebic@gmail.com>
+Date: Wed, 21 Apr 2021 23:55:51 -0700
+Subject: [PATCH] Custom Advancement Triggers
+
+
+diff --git a/src/main/java/io/papermc/paper/advancements/AdvancementManager.java b/src/main/java/io/papermc/paper/advancements/AdvancementManager.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..e8bc61477365d75ae62fac63b0ccc3f6d2d9651b
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/advancements/AdvancementManager.java
+@@ -0,0 +1,22 @@
++package io.papermc.paper.advancements;
++
++import com.google.common.collect.Maps;
++import org.bukkit.NamespacedKey;
++import org.jetbrains.annotations.NotNull;
++
++import java.util.Map;
++
++public interface AdvancementManager {
++    /**
++     * Registers a custom trigger
++     *
++     * @param trigger
++     * @param <T>
++     * @return
++     */
++    @NotNull
++    <T extends AdvancementTrigger> T registerCustomTrigger(@NotNull T trigger);
++
++    @NotNull
++    Map<NamespacedKey, AdvancementTrigger> getCustomTriggers();
++}
+diff --git a/src/main/java/io/papermc/paper/advancements/AdvancementTrigger.java b/src/main/java/io/papermc/paper/advancements/AdvancementTrigger.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..efec33406909cc860e55377108fa5a5c0aeb0010
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/advancements/AdvancementTrigger.java
+@@ -0,0 +1,32 @@
++package io.papermc.paper.advancements;
++
++import com.google.gson.JsonObject;
++import org.bukkit.Keyed;
++import org.bukkit.entity.Player;
++import org.jetbrains.annotations.NotNull;
++
++import java.util.function.Predicate;
++
++/**
++ * Should be extended by all classes representing custom triggers.
++ *
++ * @param <D> a data holder class that holds information parsed from the advancement JSON
++ */
++public abstract class AdvancementTrigger<D> implements Keyed {
```

Comment:
I personally don't really know if the class name makes a lot of sense here.
The "general" advancement trigger internally maintains a custom advancement trigger (which judging from the name seems to be a more specific implementation (but it isn't)).

Shouldn't this be the custom (plugin defined) advancement trigger ?

Reply:
Yeah, renamed a bit of stuff to make this `CustomAdvancementTrigger`.